src/check/parse/tokenize/src/main.zig

```diff
@@ -0,0 +1,1399 @@
+        return keywords.get(bytes);
+    }
+};
+
+/// Indent holds counts for spaces and tabs.
+pub const Indent = struct {
+    num_spaces: u16,
+    num_tabs: u16,
+
+    pub fn init() Indent {
+        return Indent{ .num_spaces = 0, .num_tabs = 0 };
+    }
+
+    /// Returns:
+    /// - Some(true) if self is indented more than the other indent,
+    /// - Some(false) if less,
+    /// - null if the comparison is undefined.
+    pub fn isIndentedMoreThan(self: Indent, other: ?Indent) ?bool {
+        if (other) |o| {
+            if (self.num_spaces == o.num_spaces) {
+                return self.num_tabs > o.num_tabs;
+            } else if (self.num_tabs == o.num_tabs) {
+                return self.num_spaces > o.num_spaces;
+            } else {
+                return null;
+            }
+        } else {
+            // No minimum indent means we’re always “more indented.”
+            return true;
+        }
+    }
+};
+
+/// A simple line record: the index in the token list and its indent.
+pub const Line = struct {
+    token_start: u32,
+    indent: Indent,
+};
+
+/// The buffer that accumulates tokens.
+pub const TokenizedBuffer = struct {
+    allocator: std.mem.Allocator,
+    tokens: std.MultiArrayList(Token),
+    lines: std.ArrayList(Line),
+
+    pub fn init(allocator: std.mem.Allocator) !TokenizedBuffer {
+        return TokenizedBuffer{
+            .allocator = allocator,
+            .tokens = std.MultiArrayList(Token){},
+            .lines = std.ArrayList(Line).init(allocator),
+        };
+    }
+
+    /// Pushes a token with the given kind, token offset, and length.
+    pub fn pushToken(self: *TokenizedBuffer, kind: T, tok_offset: usize, tok_length: usize) !void {
+        try self.tokens.append(self.allocator, .{
+            .kind = kind,
+            .offset = @intCast(tok_offset),
+            .length = @intCast(tok_length),
+        });
+    }
+
+    /// Returns the offset of the token at index `idx`.
+    pub fn offset(self: *TokenizedBuffer, idx: u32) usize {
+        return @intCast(self.offsets.items[@intCast(idx)]);
+    }
+};
+
+/// A comment is represented by its begin and end offsets.
+pub const Comment = struct {
+    begin: usize,
+    end: usize,
+};
+
+/// MessageKind enumerates different diagnostic messages.
+pub const MessageKind = enum {
+    MisplacedCarriageReturn,
+    AsciiControl,
+    LeadingZero,
+    UnknownToken,
+    OpaqueNameWithoutName,
+    UppercaseBase,
+    InvalidUnicodeEscapeSequence,
+    InvalidEscapeSequence,
+    UnclosedString,
+    UnclosedSingleQuote,
+    BadNumberSuffix,
+    OverClosedBrace,
+    MismatchedBrace,
+};
+
+const UnicodeKind = enum {
+    LetterUpper,
+    LetterNotUpper,
+    Digit,
+    Other,
+    Invalid,
+};
+
+const UnicodeInfo = struct {
+    kind: UnicodeKind,
+    length: usize,
+};
+
+/// A message with its kind and offset.
+pub const Message = struct {
+    kind: MessageKind,
+    begin: u32,
+    end: u32,
+};
```

Comment:
Nit: put `Message` with `MessageKind` in source definition.

```suggestion

/// A message with its kind and offset.
pub const Message = struct {
    kind: MessageKind,
    begin: u32,
    end: u32,
};

const UnicodeKind = enum {
    LetterUpper,
    LetterNotUpper,
    Digit,
    Other,
    Invalid,
};

const UnicodeInfo = struct {
    kind: UnicodeKind,
    length: usize,
};
```

Reply:
if you want to go full zig:

```zig
pub const Message = struct {
    kind: Kind,
    /// Offset into foobar.
    begin: u32,
    /// Offset into foobar.
    end: u32,

    pub const Kind = enum {
        // ...
    };
};
const Unicode = struct {
    kind: Kind,
    length: usize,

    const Kind = enum {
        LetterUpper,
        LetterNotUpper,
        Digit,
        Other,
        Invalid,
    };
};
```

notes:
 * ["Info" is always redundant](https://ziglang.org/documentation/master/#Avoid-Redundancy-in-Names)
 * "Tag" is a great alternative to "Kind" if you want to shave off 1 character :grinning: 
 * take advantage of namespaces